### DISCUSSION TITLE
Using the K3s bundled CNI with Canal

### DISCUSSION THREAD

**Discussion by retr0h** [k3s-io/k3s#6711] [Q&A]
K3s default CNI is Flannel.  Flannel is bundled as part of the K3s deployment.

I have successfully setup Canal on a K3s install, following the [instructions](https://projectcalico.docs.tigera.io/getting-started/kubernetes/flannel/flannel) provided by projectcalico.

While inspecting the [install manifest](https://raw.githubusercontent.com/projectcalico/calico/v3.24.5/manifests/canal.yaml), I noticed a version of flannel is deployed as part of the Calico Policy engine.

I understand this question is somewhat directed towards Canal, but given K3s uniqueness in bundling a CNI as part of the distribution, I was hoping  someone could help me understand if the Flannel shipped in the Canal deployment must always match that which K3s is bundled with, and/or any other caveats.

**Comment by brandond** [comment]
If you start k3s with `--flannel-backend=none` and deploy your own CNI, then it doesn't really matter what version we ship with. Anything that would care about that has been disabled.

**Comment by retr0h** [comment]
@brandond Yeah, Canal allows us to continue to use CNI distributed with K3s,  but deploys a pod in kube-system specifically for flannel's kube-subnet-mgr backend, and leverages the groovy NetworkPolicy.

```
command: [ "/opt/bin/flanneld", "--ip-masq", "--kube-subnet-mgr" ]
```

This was more or less the crux of my question.  Everything seems to work perfectly, by taking a default K3s install, and simply applying the [Canal manifest](https://raw.githubusercontent.com/projectcalico/calico/v3.24.5/manifests/canal.yaml), however, I was unsure of the interaction between the flannel used by Canal and the one used internally by k3s.

**Reply by brandond** [reply]
Ah I see. So you're **NOT** disabling the bundled flannel? I'm not sure how exactly that will work. If you want to use canal for CNI, then you should disable flannel. If you want to use it for policy, then you should disable the built-in kube-router network policy controller. If you're trying to run everything at the same time, I'm not sure what exactly is going to happen.

**Reply by retr0h** [reply]
Thanks for getting back to me @brandond. 

I'm primarily interested in [GlobalNetworkPolicy](https://projectcalico.docs.tigera.io/reference/resources/globalnetworkpolicy), such that I can restrict access to the Kubernetes host, as services which run on the host are exposed to the network.  (any services running with host networking, kube api, etc..).  Since kubernetes controls iptables, k3s/k8s didn't appear to have a native solution to this particular issue, as it relates specifically to the Host OS.
